text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+
+With this proposed mechanism, its very likely this code will be released without performing a version bump.
+
+In contrast, imagine the reverse, what if we push code that contains a version bump, making the framework incompatible, while in reality, existing functionality is not broken?
+
+This begs the question, when do we actually want a version bump? On any change to the `manifest.json`? Perhaps only on changes that break the CLI? Maybe we should require CLI upgrades only when detecting usages of specific features?
+
+The reason I call this *accidental*, is because this mechanism tries to shift most of the compatibility concerns away from humans, and into the hands of the machine. While this is generally a good approach, it does mean that we might be relaxing our brains a bit too much, causing some affects that possible wouldn't have happened in a more manual state of mind.
+
+## Intentional Breaking Changes
+
+What happens if we introduce a breaking change that isn't related to cx-protocol, causing the old integration tests to fail with a new version of either the CLI of Framework?
+
+The scenario essentially contradicts our initial assumption that old integration tests should never fail (assuming protocol version remains the same). However, it might still happen...
+
+We should discuss this, but my thoughts on the subject are:
+
+If a scenario like that occurs, it might mean the integration test isn't actually testing the right thing.
+It feels like integration tests should only validate functionality we shouldn't be breaking (cdk command line options, docker build command, and other user facing behavior).
+
+There are definitely cases where we "break" some behavior, that isn't necessarily impacting users that much, which is why its ok to do.
+
+For example, the [simplify docker asset publishing](https://github.com/aws/aws-cdk/pull/5733) PR removed the ECRAdoptedRepository construct, which is technically a breaking change. The failing tests due to that, are unit tests, not integration tests (need to double check). My argument is that if we did have an integration test that validates the existence of an `ECRAdoptedRepository`, that validation probably didn't belong their in the first place. In essence, it means the integration test was validating the implementation rather than the expected result, which is that images are properly treated as assets (uploaded to somewhere you can download it from).
+
+#### Caveat
+
+Even if we do agree that such a scenario means the test should be changed, how would we actually escape the failing test? We cannot change the previous test...
```

Comment:
Solution: always using the latest compatibility matrix? That way we can retroactively disable tests.

Reply:
Not sure what you mean :\

Should our integration tests have a "disabled" label, so we are able to externally disable the tests without touching its code?

Reply:
Not necessarily at the level of enable/disable, but the level of "this test should succeed/fail against version combinations X+Y, Z+W" etc. Of course, that requires externalizing the compatibility information from a chain of `if/elifs` that lives inside the test to something data-driven that lives outside the test itself.

Reply:
Yeap, this makes sense. Its basically providing the functionality I specified in `test-framework-compatibility.sh`, but at the granularity of a specific test, as opposed to the entire suite, in a data driven way. 👍